xtask/bench/src/utils.rs

```
@@ -0,0 +1,46 @@
+use crate::err_to_string;
+use ansi_rgb::{red, Foreground};
+use std::path::PathBuf;
+use std::str::FromStr;
+
+pub fn get_code(lib: &str) -> Result<(String, String), String> {
+    let url = url::Url::from_str(lib).map_err(err_to_string)?;
+    let segments = url
+        .path_segments()
+        .ok_or_else(|| "lib url has no segments".to_string())?;
+    let filename = segments
+        .last()
+        .ok_or_else(|| "lib url has no segments".to_string())?;
+
+    let mut file = PathBuf::from_str("target").map_err(err_to_string)?;
+    file.push(filename);
+
+    match std::fs::read_to_string(&file) {
+        Ok(code) => {
+            println!("[{}] - using [{}]", filename.fg(red()), file.display());
+            Ok((filename.to_string(), code))
+        }
+        Err(_) => {
+            println!(
+                "[{}] - Downloading [{}] to [{}]",
+                filename,
+                lib,
+                file.display()
+            );
+            match ureq::get(lib).call() {
+                Ok(response) => {
+                    let mut reader = response.into_reader();
+
+                    let _ = std::fs::remove_file(&file);
+                    let mut writer = std::fs::File::create(&file).map_err(err_to_string)?;
+                    let _ = std::io::copy(&mut reader, &mut writer);
```

Do we really want to ignore this error ? I haven't read the specific implementation of ureq so maybe it buffers the entire response body before returning but if not then the connection might fail as data is being streamed into the file

---

It should at least propagate the error like the other IO calls: `std::io::copy(&mut reader, &mut writer).map_err(err_to_string)?`, but the file could still end up being halfway written before failing. Removing the file on error would be better, something like:
```rust
if let Err(err) = std::io::copy(&mut reader, &mut writer) {
    drop(writer);
    std::fs::remove_file(&file).ok();
    return Err(err_to_string(err));
}
```